components/feature/search/src/main/java/mozilla/components/feature/search/region/RegionManager.kt

```diff
@@ -0,0 +1,123 @@
+private const val UPDATE_INTERVAL = 60 * 60 * 24 * 14
+
+// The maximum number of times we retry fetching the region from
+// the location service until we give up. We will try again on
+// the next app start.
+private const val MAX_RETRIES = 3
+
+// Timeout until we try to fetch the region again after a failure.
+private const val RETRY_TIMEOUT: Long = 10 * 60 * 1000
+
+private const val PREFERENCE_FILE = "mozac_feature_search_region"
+private const val PREFERENCE_KEY_HOME_REGION = "region.home"
+private const val PREFERENCE_KEY_CURRENT_REGION = "region.current"
+private const val PREFERENCE_KEY_REGION_FIRST_SEEN = "region.first_seen"
+
+/**
+ * Internal RegionManager for keeping track of the "current" and "home" region of a user. Used by
+ * [RegionMiddleware].
+ */
+internal class RegionManager(
+    context: Context,
+    private val locationService: LocationService,
+    @VisibleForTesting private val currentTime: () -> Long = { System.currentTimeMillis() },
+    @VisibleForTesting private val preferences: SharedPreferences = context.getSharedPreferences(
+        PREFERENCE_FILE,
+        Context.MODE_PRIVATE
+    )
+) {
+    private var homeRegion: String?
+        get() = preferences.getString(PREFERENCE_KEY_HOME_REGION, null)
+        set(value) = preferences.edit().putString(PREFERENCE_KEY_HOME_REGION, value).apply()
+
+    private var currentRegion: String?
+        get() = preferences.getString(PREFERENCE_KEY_CURRENT_REGION, null)
+        set(value) = preferences.edit().putString(PREFERENCE_KEY_CURRENT_REGION, value).apply()
+
+    private var firstSeen: Long?
+        get() = preferences.getLong(PREFERENCE_KEY_REGION_FIRST_SEEN, 0)
+        set(value) = if (value == null) {
+            preferences.edit().remove(PREFERENCE_KEY_REGION_FIRST_SEEN).apply()
+        } else {
+            preferences.edit().putLong(PREFERENCE_KEY_REGION_FIRST_SEEN, value).apply()
+        }
+
+    fun region(): RegionState? {
+        return homeRegion?.let { region ->
+            RegionState(
+                region,
+                currentRegion ?: region
+            )
+        }
+    }
+
+    suspend fun update(): RegionState? {
+        val region = fetchRegionWithRetry()?.countryCode ?: return null
+
+        if (homeRegion == null) {
+            // If we do not have a value for the home region yet, then we can set it immediately.
+            homeRegion = region
+            return RegionState(home = region, current = region)
+        }
+
+        return when (region) {
+            homeRegion -> {
+                // If we are in the home region (again) then we can clear a previously seen different
+                // "current" region.
+                currentRegion = null
+                firstSeen = null
+                null
+            }
+
+            currentRegion -> {
+                val now = currentTime()
+                if (now > (firstSeen ?: 0) + UPDATE_INTERVAL) {
```

Comment:
Looks like now/firstSeen/currentTime are in MS, but UPDATE_INTERVAL is in S right now.